### PR TITLE
Add HTTP listener service input plugin

### PR DESCRIPTION
**Pull request by bagelswitch**
The HTTP listener is a service input plugin that listens for messages sent via HTTP POST.
The plugin expects messages in the InfluxDB line-protocol ONLY, other Telegraf input data formats are not supported.
The intent of the plugin is to allow Telegraf to serve as a proxy/router for the /write endpoint of the InfluxDB HTTP API.

Example usage:  

````
curl -i -X POST 'http://localhost:8186/write' --data-binary 'cpu_load_short,host=server01,region=us-west value=0.64 1434055562000000000'
````

### Required for all PRs:

- [x] CHANGELOG.md updated
- [X] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)
- [X] README.md updated (if adding a new plugin)

### CODE REVIEW

File: Godeps [hunk 1]
```diff
@@ -25,6 +25,7 @@ github.com/gorilla/mux c9e326e2bdec29039a3761c07bece13133863e1e
 github.com/hailocab/go-hostpool e80d13ce29ede4452c43dea11e79b9bc8a15b478
 github.com/hashicorp/consul 5aa90455ce78d4d41578bafc86305e6e6b28d7d2
 github.com/hpcloud/tail b2940955ab8b26e19d43a43c4da0475dd81bdb56
+github.com/hydrogen18/stoppableListener dadc9ccc400c712e5a316107a5c462863919e579
```
Comment (sparrc) [inline Godeps]: this is only ~40 lines of code, you should just copy it in-repo instead of as a dependency

Comment (bagelswitch) [inline Godeps]: Make sense, done!

File: plugins/inputs/http_listener/http_listener.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+			var metrics []telegraf.Metric
+			metrics, err = t.parser.Parse(body)
+			if err == nil {
+				t.storeMetrics(metrics)
+			} else {
+				log.Printf("Problem parsing body: [%s], Error: %s\n", string(body), err)
+				res.WriteHeader(500)
+				res.Write([]byte("ERROR parsing metrics"))
+			}
+			res.WriteHeader(204)
```
Comment (sparrc) [inline plugins/inputs/http_listener/http_listener.go]: for these numbers you should use the `http` package constants, `http.StatusNoContent`, `http.StatusOK`, etc.

Comment (bagelswitch) [inline plugins/inputs/http_listener/http_listener.go]: Fixed.

**Comment by bagelswitch** [comment]
@kaey, in our use case the telegraf agent is already co-located with, for example, a containerized application instance in a PaaS - it sits either w/in the instance container, or on the container host, and we use multiple collectors to gather host, middleware, and application metrics from the instance - with this plugin, influx metrics emitters w/in the applications are also just pointed at the local listener - so we get a single stream of metrics covering the complete stack for this application instance that we can direct somewhere.

So in our view, this is an agent/forward-proxy use case, not an HA/reverse-proxy case where influxdb-relay would be appropriate (in fact we will use both, pointing these telegraf instances at balanced influxdb-relay instances for HA).

File: plugins/inputs/http_listener/http_listener.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	t.listener.Stop()
+	t.listener.Close()
+
+	log.Println("Stopped HTTP listener service on ", t.ServiceAddress)
+}
+
+// httpListen listens for HTTP requests.
+func (t *HttpListener) httpListen() error {
+
+	readTimeout, err := strconv.ParseInt(t.ReadTimeout, 10, 32)
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: Error not checked

File: plugins/inputs/http_listener/http_listener.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+	writeTimeout, err := strconv.ParseInt(t.WriteTimeout, 10, 32)
+
+	var server = http.Server{
+		Handler:      t,
+		ReadTimeout:  time.Duration(readTimeout) * time.Second,
+		WriteTimeout: time.Duration(writeTimeout) * time.Second,
+	}
+
+	err = server.Serve(t.listener)
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: return server.Serve(t.listener)

File: plugins/inputs/http_listener/http_listener.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+	return err
+}
+
+func (t *HttpListener) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	body, err := ioutil.ReadAll(req.Body)
+
+	if err == nil {
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: Make it err != nil and unident usual (non-error) path

File: plugins/inputs/http_listener/http_listener.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+		var path = req.URL.Path[1:]
+
+		if path == "write" {
+			var metrics []telegraf.Metric
+			metrics, err = t.parser.Parse(body)
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: Can't find where parser is initialized

Comment (bagelswitch) [inline plugins/inputs/http_listener/http_listener.go]: t.parser is initialized in HttpListener.SetParser() - I copied this pattern from the existing TCP listener plugin.

**Comment by bagelswitch** [comment]
Addressed your comments, @kaey

File: plugins/inputs/http_listener/http_listener.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/http_listener/stoppableListener"
+	"github.com/influxdata/telegraf/plugins/parsers"
+)
+
+type HttpListener struct {
+	ServiceAddress string
+	ReadTimeout    string
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: Type of timeout fields should be time.Duration

File: plugins/inputs/http_listener/http_listener.go [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+
+	return server.Serve(t.listener)
+}
+
+func (t *HttpListener) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	body, err := ioutil.ReadAll(req.Body)
+
+	if err != nil {
+		log.Printf("Problem reading request: [%s], Error: %s\n", string(body), err)
+		res.WriteHeader(http.StatusInternalServerError)
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: http.Error(w. "ERROR reading request", http.StatusInternalServerError)
return

File: plugins/inputs/http_listener/http_listener.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+	var path = req.URL.Path[1:]
+
+	if path == "write" {
+		var metrics []telegraf.Metric
+		metrics, err = t.parser.Parse(body)
+		if err == nil {
+			t.storeMetrics(metrics)
+			res.WriteHeader(http.StatusNoContent)
+			res.Write([]byte(""))
```
Comment (kaey) [inline plugins/inputs/http_listener/http_listener.go]: This call does nothing. Just remove it.

**Comment by bagelswitch** [comment]
Addressed your further comments, @kaey

**Comment by jackzampolin** [comment]
I was able to quickly get this up and running to forward points from `influx_stress` through `telegraf` to `influxdb`.  

FWIW I was able to get throughput on the plugin of 65k points a second locally while breaking the 10k point batches I was sending it into 1k point batches for writing to influx. For passing through 10k point batches throughput was closer to 80k points a second. 

When writing 10 point batches and outputting 10k point batches, more the design point, testing locally the plugin was easily able to handle 4k requests a second. It also pegged 2 of my CPUs.

I think this looks good! LGTM 👍

**Comment by kaey** [comment]
LGTM

**Comment by superbool** [comment]
why not support json format and when it will be published ? thank you !

**Comment by bagelswitch** [comment]
@superbool - I didn't think it would make sense to support other formats in this particular input, as it's replicating the InfluxDB write endpoint, which only supports the line protocol.

**Comment by kostasb** [comment]
A note that needs to go into documentation for this plugin when/if merged:

```
When chaining Telegraf instances, CREATE DATABASE requests receive 200 OK with message body `{"results":[]}` but they are not relayed. The edge Telegraf instance, meaning the one that will utterly submit data to InfluxDB, determines the destination database.
```

The PR currently needs a re-base due to new commits in the master branch.

**Comment by swang-pro** [comment]
Great news ! When will this PR be integrated in an official release ? 
Thank you

**Comment by harlow** [comment]
+1 on getting this into next release

**Comment by ddes** [comment]
+1, looks good after some tests. Thanks for your work, I really need it.

**Comment by maccman** [comment]
👍  - would love to see this merged

**Comment by sparrc** [comment]
@bagelswitch just an FYI that I've made a couple test additions, removed the locks around adding metrics (I had previously done this for the udp and tcp listeners as well), and squashed commits

this PR will be closed once I've merged that (will be done by the end of today) #1715

**Pull request closed without merging**